### PR TITLE
Change OreDictionary over from "null" to the new ItemStack system. This fixes...

### PR DESCRIPTION
**Pull request by Larsg310**
… the OreDictionary crafting replacements e.g. crafting sticks.

### CODE REVIEW

**Comment by CoderAtParadise** [comment]
You should change the itemstack arrays to nonnull lists and annotate itemstack parameters of some functions as nonnull to follow the way everything else is.

**Comment by Larsg310** [comment]
Did that, any more suggestions?

File: src/main/java/net/minecraftforge/oredict/OreDictionary.java [hunk 1]
```diff
@@ -568,7 +569,7 @@ public static boolean containsMatch(boolean strict, List<ItemStack> inputs, Item
 
     public static boolean itemMatches(ItemStack target, ItemStack input, boolean strict)
```
Comment (CoderAtParadise) [inline src/main/java/net/minecraftforge/oredict/OreDictionary.java]: Add the nonnull annotation to the ItemStacks

**Review by CoderAtParadise — CHANGES_REQUESTED** [review]

**Comment by LatvianModder** [comment]
Why stop here? The whole OreDictionary needs to be rewritten. Ideally, with no registry - just `Item#getOreNames(ItemStack stack, Collection<String> oreNames)` (or return a list/hashset, doesnt matter. I prefer HashSet, because its faster when you only need to check if it contains something). Recipes (most important thing) can check if item has that ore name, item filters and what not can do the same

**Comment by AEnterprise** [comment]
why rewrite it without registry? if there is no registry mods won't be able to add oredict tags to vanilla items

File: src/main/java/net/minecraftforge/oredict/OreDictionary.java [hunk 2]
```diff
@@ -60,7 +60,7 @@
     private static List<String>          idToName = new ArrayList<String>();
     private static Map<String, Integer>  nameToId = new HashMap<String, Integer>(128);
     private static List<NonNullList<ItemStack>> idToStack = Lists.newArrayList();
-    private static List<NonNullList<ItemStack>> idToStackUn = Lists.newArrayList();
+    private static List<List<ItemStack>> idToStackUn = Lists.newArrayList();
```
Comment (Larsg310) [inline src/main/java/net/minecraftforge/oredict/OreDictionary.java]: Undid this, with a NonNullList you can't craft ore recipes

**Comment by Larsg310** [comment]
So basically if you made the idToStackUn a NonNullList, and fix the errors related, you can't craft ore recipes... this has to get looked into

**Comment by Larsg310** [comment]
So this should all be converted and working now, with a few more commits than I'd like, but o well... Do I need to make anymore changes?

**Comment by HenryLoenwind** [comment]
If the oredict systems needs a change then it's one or both of these:
- add a flagging to the entries to mark them as "can be used as alternative" or "is equivalent and can be exchanged freely". (e.g. "Copper Ingot from mod A" for ingotCopper is the second type and "tofu" for itemFish is the first)
- remember (for the world) which was the first item to be registered (as equivalent) for one oredict name (this would allow all mods to always produce the same kind of e.g. copper)

**Pull request merged**